File: leetCode/C++/100-sameTree.cpp
#include <iostream>
#include <queue>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution1 {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (!p && !q)
            return true;
        else if (p && q && p->val == q->val)
            return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
        else
            return false;
    }
};

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        queue<TreeNode*> pqueue;
        queue<TreeNode*> qqueue;

        pqueue.push(p);
        qqueue.push(q);

        while (!pqueue.empty() && !qqueue.empty()) {
            TreeNode *p = pqueue.front();
            TreeNode *q = qqueue.front();

            pqueue.pop();
            qqueue.pop();

            if (!p && !q)
                continue;
            else if (p && q && (p->val == q->val)) {
                pqueue.push(p->left);
                qqueue.push(q->left);
                pqueue.push(p->right);
                qqueue.push(q->right);
            }
            else
                return false;
        }
        return pqueue.empty() && qqueue.empty();
    }
};

int main(int argc, char *argv []){
    Solution s;
    TreeNode t1 = TreeNode(0);
    TreeNode t2 = TreeNode(1);
    std::cout << s.isSameTree(&t1, &t2) << std::endl;
};

File: leetCode/Python2/100-sameTree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def isSameTreeRec(self, p, q): # recursive
        if p is None and q is None:
            return True
        if p and q and p.val == q.val:
            return self.isSameTree(p.left, q.left) and                         \
                   self.isSameTree(p.right, q.right)
        return False

    def isSameTree(self, p, q): # iterative
        stack = []
        stack.append([p, q])
        while stack:
            p, q = stack.pop()
            if p is not None and q is None:
                return False
            elif p is None and q is not None:
                return False
            elif p and q and p.val != q.val:
                return False
            else:
                if p is None and q is None:
                    continue
                stack.append([p.left, q.left])
                stack.append([p.right, q.right])
        return True

if __name__ == "__main__":
    s = Solution()
    t1 = TreeNode(1)
    t1.left = TreeNode(2)
    t1.right = TreeNode(3)
    t2 = TreeNode(1)
    t2.left = TreeNode(2)
    t2.right = TreeNode(3)
    print s.isSameTree(t1, t2)